tutorials/create-cloud-build-image-factory-using-packer/windows/cloudbuild.yaml

```diff
@@ -0,0 +1,41 @@
+# Perform a Packer build based on the `packer.json` configuration. This Packer
+# build creates a GCE image.
+#
+# See README.md for invocation instructions.
+steps:
+  # Retrieving secrets and placing into temporary file to reuse in later steps
+  - name: 'gcr.io/cloud-builders/gcloud'
+    entrypoint: 'bash'
+    args:
+    - '-c'
+    - |
+      echo "Retrieving secrets and placing into temporary file to reuse in later steps.." \
+      && gcloud secrets versions access latest --secret=image_factory-image_family > image_factory-image_family.txt \
+      && gcloud secrets versions access latest --secret=image_factory-image_name > image_factory-image_name.txt \
+      && gcloud secrets versions access latest --secret=image_factory-machine_type > image_factory-machine_type.txt \
+      && gcloud secrets versions access latest --secret=image_factory-region > image_factory-region.txt \
+      && gcloud secrets versions access latest --secret=image_factory-zone > image_factory-zone.txt \
+      && gcloud secrets versions access latest --secret=image_factory-network > image_factory-network.txt \
+      && gcloud secrets versions access latest --secret=image_factory-tags > image_factory-tags.txt
+
+  # Inject secret into packer template file and invoke packer build
+  - name: 'gcr.io/$PROJECT_ID/packer'
+    entrypoint: 'bash'
+    args:
+    - '-c'
+    - |
+      echo "Injecting secrets into packer template file.." \
+      && sed -i "s/_PROJECT_ID/$PROJECT_ID/g" packer.json \
+      && sed -i "s/_IMAGE_FAMILY/$(cat image_factory-image_family.txt)/g" packer.json \
+      && sed -i "s/_IMAGE_NAME/$(cat image_factory-image_name.txt)/g" packer.json \
+      && sed -i "s/_MACHINE_TYPE/$(cat image_factory-machine_type.txt)/g" packer.json \
+      && sed -i "s/_REGION/$(cat image_factory-region.txt)/g" packer.json \
+      && sed -i "s/_ZONE/$(cat image_factory-zone.txt)/g" packer.json \
+      && sed -i "s/_NETWORK/$(cat image_factory-network.txt)/g" packer.json \
+      && sed -i "s/_TAGS/$(cat image_factory-tags.txt)/g" packer.json \
+      && echo "Invoking packer build.." \
+      && packer build -debug -var project_id=$PROJECT_ID packer.json
```

Comment:
Could you use `-var` for all substitutions so that you do not need `sed` at all?

Reply:
@saltysoup , could address this suggestion?

Reply:
Cloud Build [substitutions](https://cloud.google.com/cloud-build/docs/configuring-builds/substitute-variable-values?hl=zh-TW) doesn't allow a way to modify contents of a file (packer.json) as I'm doing to inject the secrets to configure packer; I could alternatively use CB substitutions to set an environment variable (which could be pulled by packer.json), but chose not to use this method as it goes against best practices in storing secrets in env var, where it can be exposed to anybody with access to Cloud Build or external to the docker image

Reply:
What I was thinking of is something like:

```
packer build -debug \
  -var project_id=$PROJECT_ID \
  -var image_family=$(cat image_factory-image_family.txt) \
  -var image_name=$(cat image_factory-image_name.txt) \
  ... \
   packer.json
```

That way, you'd still inject all the parameters, but you do not need to use sed at all.

But your approach works too.